zone/bot_raid.cpp

```diff
@@ -0,0 +1,287 @@
+					if (strcmp(b_member->GetName(), bot_iter->GetName()) == 0) {
+						bot_iter->SetFollowID(owner);
+					} else {
+						Bot::AddBotToGroup(b_member, group_inst);
+					}
+					Bot::RemoveBotFromRaid(b_member);
+				}
+			}
+		} else {
+			Bot::RemoveBotFromRaid(bot_iter);
+		}
+	}
+}
+
+uint8 Bot::GetNumberNeedingHealedInRaidGroup(uint8& need_healed, uint8 hpr, bool includePets, Raid* raid) {
+
+	if (raid) {
+		uint32 r_group = raid->GetGroup(GetName());
+		auto raid_group_members = raid->GetRaidGroupMembers(r_group);
+
+		for (auto& m: raid_group_members) {
+			if (m.member && !m.member->qglobal) {
+				if (m.member->GetHPRatio() <= hpr) {
+					need_healed++;
+				}
+
+				if (includePets) {
+					if (m.member->GetPet() && m.member->GetPet()->GetHPRatio() <= hpr) {
+						need_healed++;
+					}
+				}
+			}
+		}
+	}
+	return need_healed;
+}
+
+void Bot::ProcessRaidInvite(Mob* invitee, Client* invitor, bool group_invite) {
+
+	if (!invitee || !invitor) {
+		return;
+	}
+
+	if (invitee->IsBot() &&
+		invitor->HasRaid() &&
+		invitee->GetOwner()->HasRaid() &&
+		invitor->GetRaid()->IsRaidMember(invitee->GetOwner()->GetName())
+	) {
+		// If the Bot Owner is in our raid we need to be able to invite their Bots
+	}
+	else if (invitee->IsBot() && (invitee->CastToBot()->GetBotOwnerCharacterID() != invitor->CharacterID())) {
+		invitor->Message(
+			Chat::Red,
+			fmt::format(
+				"{} is not your Bot. You can only invite your own Bots, or Bots that belong to a Raid member.",
+				invitee->GetCleanName()
+			).c_str()
+		);
+		return;
+	}
+
+	Raid* raid = entity_list.GetRaidByClient(invitor);
+
+	Bot::CreateBotRaid(invitee, invitor, group_invite, raid);
+}
+
+void Bot::CreateBotRaid(Mob* invitee, Client* invitor, bool group_invite, Raid* raid) {
+
+	Group* g_invitee = invitee->GetGroup();
+	Group* g_invitor = invitor->GetGroup();
+
+	if (g_invitee && invitor->IsClient()) {
+		if (!g_invitee->IsLeader(invitee)) {
+			invitor->Message(
+				Chat::Red,
+				fmt::format(
+					"You can only invite group leaders or ungrouped bots. Try {} instead.",
+					g_invitee->GetLeader()->GetCleanName()
+				).c_str()
+			);
+			return;
+		}
+	}
+
+	bool new_raid = false;
+	if (!raid) {
+		new_raid = true;
+		raid = new Raid(invitor);
+		entity_list.AddRaid(raid);
+		raid->SetRaidDetails();
+	}
+
+	// Add Invitor to new raid
+	if (new_raid) {
+		if (g_invitor) {
+			ProcessBotGroupAdd(g_invitor, raid, nullptr, true, true);
+		} else {
+			raid->SendRaidCreate(invitor);
+			raid->AddMember(invitor, 0xFFFFFFFF, true, false, true);
+			raid->SendMakeLeaderPacketTo(invitor->GetName(), invitor);
+			if (raid->IsLocked()) {
+				raid->SendRaidLockTo(invitor);
+			}
+		}
+	}
+
+	// Add Bot Group or Client Bot Group to raid
+	if (g_invitee) {
+		ProcessBotGroupAdd(g_invitee, raid);
+
+	// Add individual client to raid
+	} else if (invitee->IsClient()) {
+		ProcessBotGroupAdd(g_invitee, raid, invitee->CastToClient());
+
+	// Add individual bot to raid
+	} else {
```

Comment:
In many places of this PR. The `else` can be eliminated with early returns in the previous checks. If there are no early returns than the code falls through and we don't need to nest code.